### PR TITLE
Make DataFrame API consuming (#4621)

### LINKED ISSUE #4621: Make DataFrame Consuming
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Currently all of the various methods on `DataFrame` take `&self`, and a number of methods return `Arc<DataFrame>`.

This results in a non-trivial amount of unnecessary clones, in particular of `LogicalPlan` which is not especially cheap.

**Describe the solution you'd like**

I would like methods to return `DataFrame` instead of `Arc<DataFrame>` and for the member functions on `DataFrame` to capture `self`

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #4621
Part of #4349 
Relates to #4617 

# Rationale for this change

* Avoids unnecessary clones and memory allocations
* More idiomatic
* Fits into a broader story to rely less on interior mutability (#4617)

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/execution/context.rs [hunk 1]
```diff
@@ -265,20 +265,18 @@ impl SessionContext {
                 if_not_exists,
                 or_replace,
             }) => {
+                let input = Arc::try_unwrap(input).unwrap_or_else(|e| e.as_ref().clone());
```
Comment (tustvold) [inline datafusion/core/src/execution/context.rs:268]: This is kind of unfortunate, I hope to rework this all as part of #4617

File: datafusion/core/src/datasource/view.rs [hunk 2]
```diff
@@ -431,7 +431,7 @@ mod tests {
         )
         .await?;
 
-        ctx.register_table("t1", ctx.table("test")?)?;
+        ctx.register_table("t1", Arc::new(ctx.table("test")?))?;
```
Comment (tustvold) [inline datafusion/core/src/datasource/view.rs:434]: This API is problematic as it leads to a reference cycle (#2659)

**Comment by tustvold** [comment]
FYI @andygrove I'm not sure if this has some implication for the python bindings?

**Review by andygrove — APPROVED** [review]
LGTM

**Comment by andygrove** [comment]
> FYI @andygrove I'm not sure if this has some implication for the python bindings?

That's a good point .. thanks, I'll take a look (cc @isidentical)

**Comment by isidentical** [comment]
I think there a couple of places where we need to make some adjustments (and another interesting idea that comes with this is whether we want tests with nightly datafusion revisions on the datafusion-python) but (at least by a very light look) I don't think this should be a problem.

**Comment by tustvold** [comment]
Thank you for taking a look :+1:

**Comment by tustvold** [comment]
I'm going to merge this in as I don't think it all that controversial and there are changes backed up on it, happy to address feedback in follow up PRs

**Pull request merged**